src/cursor/cur_std.c

```diff
@@ -1175,11 +1177,77 @@ __wt_cursor_bound(WT_CURSOR *cursor, const char *config)
         WT_ERR(__wt_config_gets(session, cfg, "bound", &cval));
         if (cval.len == 0)
             WT_ERR_MSG(session, EINVAL, "setting bounds must require the bound configuration set");
+
+        /* The cursor must have a key set to place the lower or upper bound. */
+        if (!F_ISSET(cursor, WT_CURSTD_KEY_SET))
+            WT_ERR_MSG(session, EINVAL, "setting bounds must require the key to be set");
+
+        if (WT_STRING_MATCH("upper", cval.str, cval.len)) {
+            /*
+             * If the lower bounds are set, make sure that the upper bound is greater than the lower
+             * bound.
+             */
+            WT_ERR(__wt_cursor_get_raw_key(cursor, &key));
+            if (F_ISSET(cursor, WT_CURSTD_BOUND_LOWER | WT_CURSTD_BOUND_LOWER_INCLUSIVE)) {
+                WT_ERR(__wt_compare(
+                  session, CUR2BT(cursor)->collator, &key, &cursor->lower_bound, &exact));
+                if (exact < 0)
+                    WT_ERR_MSG(session, EINVAL,
+                      "The upper bounds is not lexicographically greater than the lower bound");
+            }
+            F_SET(cursor, WT_CURSTD_BOUND_UPPER | WT_CURSTD_BOUND_UPPER_INCLUSIVE);
+            WT_ERR(__wt_buf_set(session, &cursor->upper_bound, key.data, key.size));
+
+            /* Clear the key the cursor was holding. */
+            WT_CLEAR(cursor->key);
```

Comment:
I think this WT_CLEAR may leak memory.  If you really want to get rid of the key's allocated memory, you can call     `__wt_buf_free(session, &cursor->key);` and then clear.  But is there a reason the clear the key?  Does the doc say that we do?

Reply:
Brief looking at the technical docs it doesn't say that we clear the key when we set bounds. Talking with @luke-pearson we have decided free the key. I can see benefits of both methods here. If we clear the key, it provides a clear understanding next(), prev() or search() calls will behave as appriopately. I am talking about the case that if the user provides an upper bound, then performs a next(). If we do not clear the key, naturally the cursor will return WT_NOTFOUND, i am not sure if this can cause confusion to the user. @luke-pearson  what are your thoughts here?

Reply:
I don't think we want to call WT_CLEAR here. the key passed in by the user is being copied by us, we should clear the cursor key but there should a better mechanism to it.

Reply:
You know what I don't think we need to clear at all, as @ddanderson said.